zerver/lib/narrow.py

```diff
@@ -1276,6 +1297,25 @@ def fetch_messages(
     )
 
     with get_sqlalchemy_connection() as sa_conn:
+        if anchor_date:
+            # The anchor_date value needs to be parsed here because the date query can
+            # only be applied after all the other narrow conditions have been applied.
+            date_query = (
+                query.where(literal_column("zerver_message.date_sent") >= literal(anchor_date))
+                .order_by(
+                    extract(
+                        "epoch", literal_column("zerver_message.date_sent") - literal(anchor_date)
+                    )
```

Comment:
What does this `extract` part do?

Reply:
Uncommenting that part of the code produced so many logs :') I used another hack to generate the db query log:
```
SELECT id AS message_id, subject, rendered_content, pgroonga_match_positions_character(rendered_content, pgroonga_query_extract_keywords(escape_html(%(escape_html_1)s))) AS content_matches, pgroonga_match_positions_character(escape_html(subject), pgroonga_query_extract_keywords(escape_html(%(escape_html_1)s))) AS topic_matches
FROM zerver_message
WHERE recipient_id = %(recipient_id_1)s AND (search_pgroonga &@~ escape_html(%(escape_html_1)s)) AND zerver_message.date_sent >= %(param_1)s
ORDER BY EXTRACT(epoch FROM zerver_message.date_sent - %(param_2)s)
LIMIT %(param_3)s
```

with the following parameters:
```
{
  'escape_html_1': 'foo',
  'recipient_id_1': 30,
  'param_1': datetime.datetime(2023, 5, 22, 0, 0),
  'param_2': datetime.datetime(2023, 5, 22, 0, 0),
  'param_3': 1
}
```

On running `EXPLAIN ANALYZE` I got:
```
                                                                         QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=4.05..4.05 rows=1 width=342) (actual time=2.881..2.882 rows=0 loops=1)
   ->  Sort  (cost=4.05..4.05 rows=1 width=342) (actual time=2.880..2.881 rows=0 loops=1)
         Sort Key: (date_part('epoch'::text, (date_sent - '2023-05-22 00:00:00+00'::timestamp with time zone)))
         Sort Method: quicksort  Memory: 25kB
         ->  Index Scan using zerver_message_search_pgroonga on zerver_message  (cost=0.00..4.04 rows=1 width=342) (actual time=2.832..2.832 rows=0 loops=1)
               Index Cond: (search_pgroonga &@~ 'foo'::text)
               Filter: ((date_sent >= '2023-05-22 00:00:00+00'::timestamp with time zone) AND (recipient_id = 30))
               Rows Removed by Filter: 1
 Planning Time: 279.913 ms
 Execution Time: 3.067 ms
(10 rows)
```

Reply:
The `extract("epoch", literal_column("zerver_message.date_sent") - literal(anchor_date))` expression calculates the difference between the two timestamps and extracts it as an epoch value. By ordering the query based on this epoch difference, the closest message to the `anchor_date` will be selected.

Reply:
I've added a comment to explain the use `extract` function:
```.py
# Order the query results based on the absolute difference between the
# `date_sent` column and `anchor_date`. The `extract` function converts
# the timestamps to epoch values for accurate ordering.
```

Reply:
Should we call this `epoch_difference` instead?

Reply:
> Does the database have messages sorted by date already?

Mostly -- there's exceptions to that rule because of data imports. But that's true within an individual realm. 

Hmm. What I'm worried about here is that we're not taking advantage of the database index that we have on the `date_sent` column, and this operation could be very inefficient... but that index isn't very useful, since it isn't limited to a given realm or anything, so effectively what we're doing is asking the database to walk all the message matching whatever the rest of the query is, and sort those by this date filter.

One option would be to just first find the message ID in the realm that is closest to the target date via some sort of binary search mechanism -- worst case we add a new `(realm, date_sent)` index to make this step efficient.

One option would be to figure out a different query construction that we can use to ask the database to walk a new `(realm, date_sent)` index or `(recipient, date_sent)` index.